code/modules/mob/living/simple_animal/friendly/diona.dm

```diff
@@ -162,28 +159,32 @@
 		unEquip(W)
 
 	qdel(src)
+	return TRUE
 
-/mob/living/simple_animal/diona/verb/steal_blood()
-	set category = "Diona"
-	set name = "Steal Blood"
-	set desc = "Take a blood sample from a suitable donor."
-
+/mob/living/simple_animal/diona/proc/steal_blood()  //called by the hud (/obj/screen/diona/blood)
+	if(!isnymph(src) || src.stat)
+		return FALSE
+	
 	var/list/choices = list()
 	for(var/mob/living/carbon/human/H in oview(1,src))
-		choices += H
+		if(Adjacent(H))
+			choices += H
+
+	if(!choices.len)
+		to_chat(src, "<span class='warning'>No suitable blood donors nearby.</span>")
 
 	var/mob/living/carbon/human/M = input(src,"Who do you wish to take a sample from?") in null|choices
 
-	if(!M || !src)
+	if(!M || !src || !(Adjacent(M)) || src.stat) //input can take a while, so re-validate
 		return
 
-	if(NO_BLOOD in M.dna.species.species_traits)
+	if(!M.dna || NO_BLOOD in M.dna.species.species_traits)
```

Comment:
Make sure to keep the `NO_BLOOD in M.dna...` thing in its own parens, to avoid operation grouping from causing errant behavior